content/en/blog/_posts/2022-11-04-live-and-let-live-with-kluctl-and-ssa.md

```diff
@@ -0,0 +1,207 @@
+
+## Live and let live
+
+One of the main philosophies that Kluctl follows is ["live and let live"](https://kluctl.io/docs/philosophy/#live-and-let-live),
+meaning that it will try its best to work in conjunction with any other tool or
+controller running outside or inside your clusters. Kluctl will not overwrite
+any fields that it lost ownership of, unless you explicitly tell it to do so.
+
+Achieving this would not have been possible (or at least several magnitudes
+harder) without the use of SSA. Server-side apply allows Kluctl
+to detect when ownership for a field got lost, for example when another controller
+or operator updates that field to another value. Kluctl can then decide on a 
+field-by-field basis if force-applying is required before retrying based on these
+decisions.
+
+## The days before SSA
+
+The first versions of Kluctl were based on shelling out to `kubectl` and thus
+implicitly relied on client-side apply. At that time, SSA was
+still alpha and quite buggy. And to be honest, I didn't even know it was a
+thing at that time.
+
+The way client-side apply worked had some serious drawbacks. The most obvious one
+(it was guaranteed that you'd stumble on this by yourself if enough time passed)
+is that it relied on an annotation (`kubectl.kubernetes.io/last-applied-configuration`)
+being added to the object, bringing in all the limitations and issues with huge
+annotation values. A good example of such issues are
+[CRDs being so large](https://github.com/prometheus-operator/prometheus-operator/issues/4439),
+that they don't fit into the annotation's value anymore.
+
+Another drawback can be seen just by looking at the name (**client**-side apply).
+Being **client** side means that each client has to provide the apply-logic on
+its own, which at that time was only properly implemented inside `kubectl`,
+making it hard to be replicated inside controllers.
+
+This added `kubectl` as a dependency (either as an executable or in the form of
+Go packages) to all controllers that wanted to leverage the apply-logic.
+
+However, even if one managed to get client-side apply running from inside a
+controller, you ended up with a solution that gave no control over how it
+worked internally. As an example, there was no way to individually decide which
+fields to overwrite in case of external changes and which ones to let go. 
+
+## Discovering SSA apply
+
+I was never happy with the solution described above and then somehow stumbled
+across [server-side apply](/docs/reference/using-api/server-side-apply/),
+which was still in beta at that time. Experimenting with it via
+`kubectl apply --server-side` revealed immediately that the true power of
+SSA can not be easily leveraged by shelling out to `kubectl`.
+
+The way SSA is implemented in `kubectl` does not allow enough
+control over conflict resolution as it can only switch between
+"not force-applying anything and erroring out" and "force-applying everything
+without showing any mercy!".
```

Comment:
@apelisse and I have talked about adding a mechanism to tell the server which fields you want to force apply, thanks for the evidence that we would have at least one user ;)